### PR TITLE
 [Access control] Private individual page

### PR DESCRIPTION
**Pull request by litvinovg**
[VIVO PR](https://github.com/vivo-project/VIVO/pull/3930)

# What does this pull request do?
Created individual access object to check access permission to display individual pages.
Created policy templates to restrict access to individual pages:
Created useConfiguration object property to provide sparql query when sparql query results tested against provided values.
Suppress display of individual by uri (has priority 5000 ) 
[template_suppress_display_individual_page_by_uri.n3](https://github.com/vivo-project/Vitro/pull/433/files#diff-f74dfaebfbb5a86a066e787bbbd410ced9060288b6017896acedd7efe35f65d1)
On individual editing form added checkboxes to suppress display of individual for roles
image
Suppress individual page display for a role by type (has priority 1500 ) 
[template_suppress_display_individual_page_by_type.n3](https://github.com/vivo-project/Vitro/pull/433/files#diff-97f08ef85ec674e5e4f8460deba476ce3a99f78adcfb5da8578d66a1a47e2247)
Suppress display of individual by type if individual is not related to self editor profile (has priority 1500 ) 
[template_suppress_display_not_related_individual_page_by_type.n3](https://github.com/vivo-project/Vitro/pull/433/files#diff-5cbf6c679b5bd77e47e1495db4bd94b5a0e1da709897e34b4ede94408685c446)
On Class Editing Form added checkboxes to suppress display of individual in this class for roles
image

Allow display of page individual (if access wasn't denied by policies with higher priority, this policy has priority 1000) [policy_allow_display_individual_page.n3](https://github.com/vivo-project/Vitro/pull/433/files#diff-373ab424cb288594fb3041305df55f5b3563091b927d68f3956f7bbe5fabdcea) 
Suppress display of properties not related to self editor profiles 
[template_suppress_display_not_related_property_by_uri.n3](https://github.com/vivo-project/Vitro/pull/433/files#diff-d2af8eb12e90ee1d5598a058c564e84c99a41d1ab70d2e95605855bea65fe202)
On property editing form page added checkbox to suppress display of property in not related individuals
image

Added tests to for new policy templates.

Removed duplicate authorization checks without statement details and graph.
Added named key components to access data sets and value sets by keys 
[named_key_components.n3](https://github.com/vivo-project/Vitro/pull/433/files#diff-d2d31e6269eed63cd99b4c1dd41c057db45e6f8445f0ade99ff874203ad75505)
Refactoring: 
Store model in AccessObject instead of AccessObjectStatement
Renamed sparql variable personUri to profileUri as it better suits related profiles and leaving it the same could cause confusion in future.
Removed uris String array parameter in PolicyLoader.getDataSetUriByKey method as it is not used anymore, use varargs for literal identifiers. Refactored all related methods.
Replaced ProximityChecker with more abstract SparqlSelectQueryResultsChecker.
Added safety check to AttributeValueSetFactory to avoid reusing not related value set returned by key. 
Improved logging in PolicyLoader.

# How to test
Policy templates are tested in automatic tests.
Build Vitro and VIVO, try using checkboxes to suppress display of individual pages and properties.

# Interested parties
@chenejac

### CODE REVIEW

**Comment by litvinovg** [comment]
@chenejac Added translations for different languages. We need reviews/corrections from native speakers.

File: home/src/main/resources/rdf/i18n/es/interface-i18n/firsttime/vitro_UiLabel.ttl [hunk 1]
```diff
@@ -6309,3 +6309,52 @@ uil-data:captcha_user_sol_invalid.Vitro
...
+        uil:hasApp       "Vitro" ;
+        uil:hasKey       "suppress_operation_for_individuals_of_this_class" ;
+        uil:hasPackage   "Vitro-languages" .
+
+uil-data:suppress_operation_for_not_related_individuals_of_this_class.Vitro
+        rdf:type         owl:NamedIndividual ;
+        rdf:type         uil:UILabel ;
+        rdfs:label       "Evitar operaciones <b>{0}</b> para páginas de objetos no relacionados que pertenecen a la clase editada"@es ;
+        uil:hasApp       "Vitro" ;
+        uil:hasKey       "suppress_operation_for_not_related_individuals_of_this_class" ;
```
Comment (brianjlowe) [inline home/src/main/resources/rdf/i18n/es/interface-i18n/firsttime/vitro_UiLabel.ttl:6358]: Perhaps change "not related individuals" to "unrelated individuals"

File: home/src/main/resources/rdf/i18n/es/interface-i18n/firsttime/vitro_UiLabel.ttl [hunk 2]
```diff
@@ -6309,3 +6309,52 @@ uil-data:captcha_user_sol_invalid.Vitro
...
+        uil:hasApp       "Vitro" ;
+        uil:hasKey       "operation_permissions_for_this_class" ;
+        uil:hasPackage   "Vitro-languages" .
+
+uil-data:suppress_operation_for_this_property_in_not_related_individuals.Vitro
+        rdf:type         owl:NamedIndividual ;
+        rdf:type         uil:UILabel ;
+        rdfs:label       "Impedir operaciones <b>{0}</b> para esta propiedad en páginas de objetos no relacionados"@es ;
+        uil:hasApp       "Vitro" ;
+        uil:hasKey       "suppress_operation_for_this_property_in_not_related_individuals" ;
```
Comment (brianjlowe) [inline home/src/main/resources/rdf/i18n/es/interface-i18n/firsttime/vitro_UiLabel.ttl:6342]: Perhaps change "not related individuals" to "unrelated individuals"

File: home/src/main/resources/rdf/i18n/en_US/interface-i18n/firsttime/vitro_UiLabel.ttl [hunk 3]
```diff
@@ -6309,3 +6309,52 @@ uil-data:captcha_user_sol_invalid.Vitro
...
+        uil:hasApp       "Vitro" ;
+        uil:hasKey       "suppress_operation_for_individuals_of_this_class" ;
+        uil:hasPackage   "Vitro-languages" .
+
+uil-data:suppress_operation_for_not_related_individuals_of_this_class.Vitro
+        rdf:type         owl:NamedIndividual ;
+        rdf:type         uil:UILabel ;
+        rdfs:label       "Suppress <b>{0}</b> for not related individual pages of this class"@en-US ;
+        uil:hasApp       "Vitro" ;
+        uil:hasKey       "suppress_operation_for_not_related_individuals_of_this_class" ;
```
Comment (brianjlowe) [inline home/src/main/resources/rdf/i18n/en_US/interface-i18n/firsttime/vitro_UiLabel.ttl:6358]: Perhaps change "not related individuals" to "unrelated individuals"

File: home/src/main/resources/rdf/i18n/en_US/interface-i18n/firsttime/vitro_UiLabel.ttl [hunk 4]
```diff
@@ -6309,3 +6309,52 @@ uil-data:captcha_user_sol_invalid.Vitro
...
+        rdf:type         uil:UILabel ;
+        rdfs:label       "<b>{0}</b> permissions for this class"@en-US ;
+        uil:hasApp       "Vitro" ;
+        uil:hasKey       "operation_permissions_for_this_class" ;
+        uil:hasPackage   "Vitro-languages" .
+
+uil-data:suppress_operation_for_this_property_in_not_related_individuals.Vitro
+        rdf:type         owl:NamedIndividual ;
+        rdf:type         uil:UILabel ;
+        rdfs:label       "Suppress <b>{0}</b> for this property in not related individuals"@en-US ;
```
Comment (brianjlowe) [inline home/src/main/resources/rdf/i18n/en_US/interface-i18n/firsttime/vitro_UiLabel.ttl:6340]: Perhaps change "not related individuals" to "unrelated individuals"

Comment (litvinovg) [reply home/src/main/resources/rdf/i18n/en_US/interface-i18n/firsttime/vitro_UiLabel.ttl]: Fixed

File: home/src/main/resources/rdf/i18n/en_US/interface-i18n/firsttime/vitro_UiLabel.ttl (hunk 3 above)
Comment (litvinovg) [reply home/src/main/resources/rdf/i18n/en_US/interface-i18n/firsttime/vitro_UiLabel.ttl]: Fixed

File: home/src/main/resources/rdf/i18n/es/interface-i18n/firsttime/vitro_UiLabel.ttl (hunk 2 above)
Comment (litvinovg) [reply home/src/main/resources/rdf/i18n/es/interface-i18n/firsttime/vitro_UiLabel.ttl]: Fixed

File: home/src/main/resources/rdf/i18n/es/interface-i18n/firsttime/vitro_UiLabel.ttl (hunk 1 above)
Comment (litvinovg) [reply home/src/main/resources/rdf/i18n/es/interface-i18n/firsttime/vitro_UiLabel.ttl]: Fixed

**Review by bkampe — APPROVED** [review]
I checked the German translations and think they are fine. Maybe a better wording would be possible, but it's not that easy to build well formed german sentences with these more or less complex statements.

File: home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl [hunk 5]
```diff
@@ -6309,3 +6309,52 @@ uil-data:captcha_user_sol_invalid.Vitro
         uil:hasApp      "Vitro" ;
         uil:hasKey      "captcha_user_sol_invalid" ;
         uil:hasPackage  "Vitro-languages" .
+
+	uil-data:suppress_operation_for_roles.Vitro
+        rdf:type         owl:NamedIndividual ;
+        rdf:type         uil:UILabel ;
+        rdfs:label       "Verweigern Sie <b>{0}</b>-Vorgänge für Rollen"@de-DE ;
```
Comment (hauschke) [inline home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl:6316]: rdfs:label       "<b>{0}</b>-Vorgänge für Rollen verweigern"@de-DE ;

File: home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl [hunk 6]
```diff
@@ -6309,3 +6309,52 @@ uil-data:captcha_user_sol_invalid.Vitro
...
+        uil:hasApp       "Vitro" ;
+        uil:hasKey       "suppress_operation_for_roles" ;
+        uil:hasPackage   "Vitro-languages" .
+
+uil-data:operation_permissions_for_this_property.Vitro
+        rdf:type         owl:NamedIndividual ;
+        rdf:type         uil:UILabel ;
+        rdfs:label       "Erlauben Sie <b>{0}</b>-Vorgänge für diese Eigenschaft"@de-DE ;
```
Comment (hauschke) [inline home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl:6324]: rdfs:label       "<b>{0}</b>-Vorgänge für diese Property erlauben"@de-DE ;

File: home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl [hunk 7]
```diff
@@ -6309,3 +6309,52 @@ uil-data:captcha_user_sol_invalid.Vitro
...
+        uil:hasApp       "Vitro" ;
+        uil:hasKey       "operation_permissions_for_this_property" ;
+        uil:hasPackage   "Vitro-languages" .
+
+uil-data:operation_permissions_for_this_class.Vitro
+        rdf:type         owl:NamedIndividual ;
+        rdf:type         uil:UILabel ;
+        rdfs:label       "Erlauben Sie <b>{0}</b>-Vorgänge für diese Klasse"@de-DE ;
```
Comment (hauschke) [inline home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl:6332]: rdfs:label       "<b>{0}</b>-Vorgänge für diese Klasse erlauben"@de-DE ;

File: home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl [hunk 8]
```diff
@@ -6309,3 +6309,52 @@ uil-data:captcha_user_sol_invalid.Vitro
...
+        uil:hasApp       "Vitro" ;
+        uil:hasKey       "operation_permissions_for_this_class" ;
+        uil:hasPackage   "Vitro-languages" .
+
+uil-data:suppress_operation_for_this_property_in_unrelated_individuals.Vitro
+        rdf:type         owl:NamedIndividual ;
+        rdf:type         uil:UILabel ;
+        rdfs:label       "Verhindern Sie <b>{0}</b>-Vorgänge für diese Eigenschaft auf nicht verwandten Objektseiten"@de-DE ;
```
Comment (hauschke) [inline home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl:6340]: rdfs:label       "<b>{0}</b>-Vorgänge für diese Property bei nicht verwandten Instanzen verhindern"@de-DE ;

File: home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl [hunk 9]
```diff
@@ -6309,3 +6309,52 @@ uil-data:captcha_user_sol_invalid.Vitro
...
+        rdf:type         uil:UILabel ;
+        rdfs:label       "Verweigern Sie <b>{0}</b>-Vorgänge für Objektseiten, die zur bearbeiteten Klasse gehören"@de-DE ;
+        uil:hasApp       "Vitro" ;
+        uil:hasKey       "suppress_operation_for_individuals_of_this_class" ;
+        uil:hasPackage   "Vitro-languages" .
+
+uil-data:suppress_operation_for_unrelated_individuals_of_this_class.Vitro
+        rdf:type         owl:NamedIndividual ;
+        rdf:type         uil:UILabel ;
+        rdfs:label       "Verhindern Sie <b>{0}</b>-Vorgänge für Seiten mit nicht verwandten Objekten, die zur bearbeiteten Klasse gehören"@de-DE ;
```
Comment (hauschke) [inline home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl:6356]: rdfs:label       "<b>{0}</b>-Vorgänge für Seiten mit nicht verwandten Instanzen, die zur bearbeiteten Klasse gehören, verhindern"@de-DE ;

File: home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl (hunk 5 above)
Comment (litvinovg) [reply home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl]: Fixed

File: home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl (hunk 6 above)
Comment (litvinovg) [reply home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl]: Fixed

File: home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl (hunk 7 above)
Comment (litvinovg) [reply home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl]: Fixed

File: home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl (hunk 8 above)
Comment (litvinovg) [reply home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl]: Fixed

File: home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl (hunk 9 above)
Comment (litvinovg) [reply home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl]: Fixed

**Comment by hauschke** [comment]
Thanks for the changes in German translation, @litvinovg, looks good now.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/policy/PolicyLoader.java [hunk 10]
```diff
@@ -478,9 +499,14 @@ protected void processQuerySolution(QuerySolution qs) {
         return values;
     }
 
-    public String getEntityValueSetUri(AccessOperation ao, AccessObjectType aot, String role) {
-        long expectedSize = 3;
-        String queryText = getDataSetByKeyQuery(new String[] { }, new String[] { ao.toString(), aot.toString(), role });
+    public String getEntityValueSetUri(AccessOperation ao, AccessObjectType aot, String role,
+            String... namedKeyComponents) {
+        int expectedSize = 3 + namedKeyComponents.length;
+        String[] ids = Arrays.copyOf(namedKeyComponents, expectedSize);
```
Comment (chenejac) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/policy/PolicyLoader.java:505]: Can you please align this with https://github.com/vivo-project/Vitro/pull/433/files#diff-2ccd8e24d26e6788b9443cf88e40fd036d2e231a5bc70baa98ebfb756ef0b7f3R101?

File: api/src/test/java/edu/cornell/mannlib/vitro/webapp/auth/policy/AllowDisplayIndividualPagePolicy.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+import java.util.Collections;
+import java.util.Set;
+
+import edu.cornell.mannlib.vitro.webapp.auth.objects.AccessObject;
+import edu.cornell.mannlib.vitro.webapp.auth.objects.IndividualAccessObject;
+import edu.cornell.mannlib.vitro.webapp.auth.objects.NamedAccessObject;
+import edu.cornell.mannlib.vitro.webapp.auth.requestedAction.SimpleAuthorizationRequest;
+import org.junit.Test;
+
+public class AllowDisplayIndividualPagePolicy extends PolicyTest {
```
Comment (chenejac) [inline api/src/test/java/edu/cornell/mannlib/vitro/webapp/auth/policy/AllowDisplayIndividualPagePolicy.java:21]: Usually Test classes contains suffix Test

File: home/src/main/resources/rdf/i18n/sr_Latn_RS/interface-i18n/firsttime/vitro_UiLabel.ttl [hunk 12]
```diff
@@ -6309,3 +6309,52 @@ uil-data:captcha_user_sol_invalid.Vitro
         uil:hasApp      "Vitro" ;
         uil:hasKey      "captcha_user_sol_invalid" ;
         uil:hasPackage  "Vitro-languages" .
+
+uil-data:suppress_operation_for_roles.Vitro
+        rdf:type         owl:NamedIndividual ;
+        rdf:type         uil:UILabel ;
+        rdfs:label       "Odbij <b>{0}</b> operacije za uloge"@sr-Latn-RS ;
```
Comment (chenejac) [inline home/src/main/resources/rdf/i18n/sr_Latn_RS/interface-i18n/firsttime/vitro_UiLabel.ttl:6316]:
```suggestion
        rdfs:label       "Zabrani <b>{0}</b> operacije za uloge"@sr-Latn-RS ;
```

File: home/src/main/resources/rdf/i18n/sr_Latn_RS/interface-i18n/firsttime/vitro_UiLabel.ttl [hunk 13]
```diff
@@ -6309,3 +6309,52 @@ uil-data:captcha_user_sol_invalid.Vitro
...
+        rdf:type         uil:UILabel ;
+        rdfs:label       "Dozvoli <b>{0}</b> operacije za ovu klasu"@sr-Latn-RS ;
+        uil:hasApp       "Vitro" ;
+        uil:hasKey       "operation_permissions_for_this_class" ;
+        uil:hasPackage   "Vitro-languages" .
+
+uil-data:suppress_operation_for_this_property_in_unrelated_individuals.Vitro
+        rdf:type         owl:NamedIndividual ;
+        rdf:type         uil:UILabel ;
+        rdfs:label       "Sprečite <b>{0}</b> operacije za ovo svojstvo na stranicama nepovezanih objekata"@sr-Latn-RS ;
```
Comment (chenejac) [inline home/src/main/resources/rdf/i18n/sr_Latn_RS/interface-i18n/firsttime/vitro_UiLabel.ttl:6340]:
```suggestion
        rdfs:label       "Zabrani <b>{0}</b> operacije za ovo svojstvo na stranicama nepovezanih objekata"@sr-Latn-RS ;
```

File: home/src/main/resources/rdf/i18n/sr_Latn_RS/interface-i18n/firsttime/vitro_UiLabel.ttl [hunk 14]
```diff
@@ -6309,3 +6309,52 @@ uil-data:captcha_user_sol_invalid.Vitro
...
+        uil:hasApp       "Vitro" ;
+        uil:hasKey       "suppress_operation_for_this_property_in_unrelated_individuals" ;
+        uil:hasPackage   "Vitro-languages" .
+
+uil-data:suppress_operation_for_individuals_of_this_class.Vitro
+        rdf:type         owl:NamedIndividual ;
+        rdf:type         uil:UILabel ;
+        rdfs:label       "Odbij <b>{0}</b> operacije za stranice objekata koje pripadaju uređenoj klasi"@sr-Latn-RS ;
```
Comment (chenejac) [inline home/src/main/resources/rdf/i18n/sr_Latn_RS/interface-i18n/firsttime/vitro_UiLabel.ttl:6348]:
```suggestion
        rdfs:label       "Zabrani <b>{0}</b> operacije za objekte koje pripadaju ovoj klasi"@sr-Latn-RS ;
```

File: home/src/main/resources/rdf/i18n/sr_Latn_RS/interface-i18n/firsttime/vitro_UiLabel.ttl [hunk 15]
```diff
@@ -6309,3 +6309,52 @@ uil-data:captcha_user_sol_invalid.Vitro
...
+        uil:hasApp       "Vitro" ;
+        uil:hasKey       "suppress_operation_for_individuals_of_this_class" ;
+        uil:hasPackage   "Vitro-languages" .
+
+uil-data:suppress_operation_for_unrelated_individuals_of_this_class.Vitro
+        rdf:type         owl:NamedIndividual ;
+        rdf:type         uil:UILabel ;
+        rdfs:label       "Sprečavanje <b>{0}</b> operacija za stranice nepovezanih objekata koji pripadaju uređenoj klasi"@sr-Latn-RS ;
```
Comment (chenejac) [inline home/src/main/resources/rdf/i18n/sr_Latn_RS/interface-i18n/firsttime/vitro_UiLabel.ttl:6356]:
```suggestion
        rdfs:label       "Zabrani <b>{0}</b> operacija za nepovezane objekte koji pripadaju ovoj klasi"@sr-Latn-RS ;
```

**Review by chenejac — CHANGES_REQUESTED** [review]
@litvinovg great work. Can you please check a couple of my comments?

File: api/src/test/java/edu/cornell/mannlib/vitro/webapp/auth/policy/AllowDisplayIndividualPagePolicy.java (hunk 11 above)
Comment (litvinovg) [reply api/src/test/java/edu/cornell/mannlib/vitro/webapp/auth/policy/AllowDisplayIndividualPagePolicy.java]: Fixed

File: api/src/main/java/edu/cornell/mannlib/vedit/controller/BaseEditController.java [hunk 16]
```diff
@@ -242,16 +243,110 @@ protected static void addAccessAttributes(HttpServletRequest req, String entityU
...
+        
+        RoleInfo role = getSelfEditorRole(req);
+        Map<String, List<RoleInfo>> propertySuppressionsToRoles = new LinkedHashMap<>();
+        List<RoleInfo> roleInfos = new LinkedList<>();
+        String operationName = StringUtils.capitalize(operation.toString().toLowerCase());
+        propertySuppressionsToRoles.put(operationName, roleInfos);
+        roleInfos.add(role);
+
+        getRolePolicyInformation(entityURI, aot, namedKeys, operation, roleInfos);
+        req.setAttribute("propertySuppressionsNotRelated", propertySuppressionsToRoles);
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vedit/controller/BaseEditController.java]: Fixed

File: api/src/main/java/edu/cornell/mannlib/vedit/controller/BaseEditController.java [hunk 17]
```diff
@@ -242,16 +243,110 @@ protected static void addAccessAttributes(HttpServletRequest req, String entityU
...
+        Map<String, List<RoleInfo>> typeSuppressionsToRoles = new LinkedHashMap<>();
+        List<RoleInfo> roleInfos = new LinkedList<>();
+        String operationName = StringUtils.capitalize(operation.toString().toLowerCase());
+        typeSuppressionsToRoles.put(operationName, roleInfos);
+        for (RoleInfo role : roles) {
+            RoleInfo roleCopy = role.clone();
+            roleInfos.add(roleCopy);
+        }
+        getRolePolicyInformation(entityURI, aot, namedKeys, operation, roleInfos);
+        req.setAttribute("typeSuppressions", typeSuppressionsToRoles);
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vedit/controller/BaseEditController.java]: Fixed

File: api/src/main/java/edu/cornell/mannlib/vedit/controller/BaseEditController.java [hunk 18]
```diff
@@ -242,16 +243,110 @@ protected static void addAccessAttributes(HttpServletRequest req, String entityU
...
             }
-            if (!StringUtils.isEmpty(entityURI)) {
-                for (RoleInfo roleInfo : roleInfos) {
-                    if (roleInfo.isEnabled()) {
-                        roleInfo.setGranted(
-                                EntityPolicyController.isGranted(entityURI, aot, operation, roleInfo.getUri()));
-                    }
+            getRolePolicyInformation(entityURI, aot, namedKeys, operation, roleInfos);
+        }
+        req.setAttribute("operationsToRoles", operationsToRoles);
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vedit/controller/BaseEditController.java]: Fixed

**Review by chenejac — DISMISSED** [review]
well done

File: home/src/main/resources/rdf/i18n/sr_Latn_RS/interface-i18n/firsttime/vitro_UiLabel.ttl [hunk 19]
```diff
@@ -6309,3 +6309,52 @@ uil-data:captcha_user_sol_invalid.Vitro
         uil:hasApp      "Vitro" ;
         uil:hasKey      "captcha_user_sol_invalid" ;
         uil:hasPackage  "Vitro-languages" .
+
+uil-data:suppress_operation_for_roles.Vitro
+        rdf:type         owl:NamedIndividual ;
+        rdf:type         uil:UILabel ;
+        rdfs:label       "Zabrani <b>{0}</b> operacije za uloge"@sr-Latn-RS ;
```
Comment (ivanmrsulja) [inline home/src/main/resources/rdf/i18n/sr_Latn_RS/interface-i18n/firsttime/vitro_UiLabel.ttl:6316]:
```suggestion
        rdfs:label       "Zabrani <b>{0}</b> operacije za role"@sr-Latn-RS ;
```
@chenejac I'm not sure about this, it is used here as a slang of some sorts, I don't know if it's an official word :smile: "uloge" just doesn't sound good to me

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/checks/SparqlSelectQueryResultsChecker.java [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+
+    private static Set<String> getSparqlSelectResults(Model model, String profileUri, String queryTemplate,
+            AuthorizationRequest ar) {
+        HashMap<String, Set<String>> queryMap = QueryResultsMapCache.get();
+        String queryMapKey = createQueryMapKey(profileUri, queryTemplate, ar);
+        if (queryMap.containsKey(queryMapKey)) {
+            return queryMap.get(queryMapKey);
+        }
+        Set<String> results = new HashSet<>();
+        ParameterizedSparqlString pss = new ParameterizedSparqlString();
```
Comment (ivanmrsulja) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/checks/SparqlSelectQueryResultsChecker.java:92]: I wouldn't use an abbreviation here, full name is maybe long but is way more readable.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/checks/SparqlSelectQueryResultsChecker.java [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+        if (StringUtils.isBlank(queryTemplate)) {
+            queryTemplate = check.getValues().getSingleValue();
+        }
+        if (StringUtils.isBlank(queryTemplate)) {
+            log.error("SparqlQueryContains template is empty");
+            return false;
+        }
+        AccessObject ao = ar.getAccessObject();
+        Model m = ao.getModel();
+        if (m == null) {
```
Comment (ivanmrsulja) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/checks/SparqlSelectQueryResultsChecker.java:42]: I would rename "m" to "model" and use !Objects.nonNull(model)

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/checks/CheckFactory.java [hunk 22]
```diff
@@ -48,9 +49,18 @@ public static Check createCheck(QuerySolution qs, AttributeValueKey dataSetKey)
                 check = null;
         }
         check.setType(CheckType.valueOf(testId));
+        setConfiguration(qs, check);
         return check;
     }
 
+    private static void setConfiguration(QuerySolution qs, Check check) {
+        RDFNode rdfNode = qs.get("config");
+        if (rdfNode == null || !rdfNode.isLiteral()) {
```
Comment (ivanmrsulja) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/checks/CheckFactory.java:58]: As I mentioned, you can switch to using Objects.nonNull() if you have time and will to do it :smile:

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/attributes/AttributeValueKey.java [hunk 23]
```diff
@@ -9,15 +13,31 @@ public class AttributeValueKey {
...
         this.aot = aot;
         this.role = role;
         this.type = type;
+        this.namedKeyComponents = new HashSet<>(Arrays.asList(namedKeyComponents));
+    }
+
+    public AttributeValueKey(AccessOperation ao, AccessObjectType aot, String role, String type,
+            Set<String> namedKeyComponents) {
+        this.ao = ao;
+        this.aot = aot;
```
Comment (ivanmrsulja) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/attributes/AttributeValueKey.java:37]: I would rename class attributes to the full name. Constructor parameters can stay as abbreviations.

**Review by ivanmrsulja — CHANGES_REQUESTED** [review]
Looks good, but I have left some comments that you may find helpfull :slightly_smiling_face:

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/attributes/AttributeValueKey.java (hunk 23 above)
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/attributes/AttributeValueKey.java]: Fixed

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/checks/SparqlSelectQueryResultsChecker.java (hunk 20 above)
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/checks/SparqlSelectQueryResultsChecker.java:92]: It is a widely used abbreviation. I would leave it this way.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/checks/CheckFactory.java (hunk 22 above)
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/checks/CheckFactory.java:58]: It is better to leave null checks consistent in the code base, not have it Objects.isNull(var) in one place and var == null in another. I would leave it as it is.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/checks/SparqlSelectQueryResultsChecker.java (hunk 20 above)
Comment (ivanmrsulja) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/checks/SparqlSelectQueryResultsChecker.java:92]: ok then, if it is widely used

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/checks/CheckFactory.java (hunk 22 above)
Comment (ivanmrsulja) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/checks/CheckFactory.java:58]: We can switch to the better practice now, and maybe in the future refactor the entire code base? What do you think?

Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/checks/CheckFactory.java:58]: I wouldn't call it a better practice. https://community.sonarsource.com/t/set-usage-of-objects-isnull-and-objects-nonnull-as-code-smells/58469

Comment (ivanmrsulja) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/auth/checks/CheckFactory.java:58]: Ok, you are right :smile: I didn't know that it is, in fact, a code smell. Sorry.

**Review by chenejac — APPROVED** [review]

**Pull request merged**